RHIDP-3878 Authenticating with RHSSO (#480)
redhat-developer/red-hat-developers-documentation-rhdh: modules/authentication/proc-enabling-authentication-with-rhsso.adoc

.. Use an existing realm, or link:https://docs.redhat.com/en/documentation/red_hat_single_sign-on/7.6/html-single/getting_started_guide/index#create-realm_[create a realm], with a distinctive **Name** such as __<my_realm>__.
Save the value for the next step:
* **RHSSO realm base URL**, such as: __<your_rhsso_URL>__/auth/realms/__<your_realm>__.

.. To register your {product-short} in RHSSO, in this realm, link:https://docs.redhat.com/en/documentation/red_hat_single_sign-on/7.6/html-single/getting_started_guide/index#registering-app_[create a Client ID], with:
... **Client ID**: A distinctive client ID, such as __<{product-very-short}>__.
... **Valid redirect URIs**: Set to the OIDC handler URL: `https://__<RHDH_URL>__/api/auth/oidc/handler/frame`.
... Navigate to the **Credentials** tab and copy the **Client secret**.

> ... Save the values for the next step:

Reviewer: ```suggestion
... Save the values of **Client ID** and **Client Secret** for the next step.
```

Author: This is less readable.

> * **Client ID**

Reviewer: ```suggestion

```

> * **Client Secret**

Reviewer: ```suggestion

```

.. To prepare for the verification steps, in the same realm, get the credential information for an existing user or link:https://docs.redhat.com/en/documentation/red_hat_single_sign-on/7.6/html-single/getting_started_guide/index#create-user_[create a user]. Save the user credential information for the verification steps.

> . To add your RHSSO credentials in your Developer Hub secrets, edit your Developer Hub secrets, such as `secrets-rhdh`, and add the following key/value pairs:

Reviewer: ```suggestion
. To add your RHSSO credentials to your {product-short} secrets, edit your {product-short} secrets, such as `secrets-rhdh`, and add the following key/value pairs:
```

+
`AUTH_OIDC_CLIENT_ID`:: Enter the saved **Client ID**.
`AUTH_OIDC_CLIENT_SECRET`:: Enter the saved **Client Secret**.
`AUTH_OIDC_METADATA_URL`:: Enter the saved **RHSSO realm base URL**.

> . To set up the RHSSO authentication provider in your {product-short} custom configuration, edit your custom Developer Hub config map, such as `app-config-rhdh`, and add following lines to the `app-config-rhdh.yaml` content:

Reviewer: ```suggestion
. To set up the RHSSO authentication provider in your {product-short} custom configuration, edit your custom {product-short} ConfigMap such as `app-config-rhdh`, and add the following lines to the `app-config-rhdh.yaml` content:
```

+
--

> .`app-config-rhdh.yaml` fragment with mandatory fields to enable authentication with RHSSO

Reviewer: ```suggestion
.Example `app-config-rhdh.yaml` fragment with mandatory fields to enable authentication with RHSSO
```

Author: For minimalism, fragment is enough, example would be a répétition.

[source,yaml]
----
auth:

>   environment: production

Reviewer: ```suggestion
  environment: production <1>
```

  providers:
    oidc:
      production:

>         metadataUrl: ${AUTH_OIDC_METADATA_URL}

Reviewer: ```suggestion
        metadataUrl: ${AUTH_OIDC_METADATA_URL} <2>
```

>         clientId: ${AUTH_OIDC_CLIENT_ID}

Reviewer: ```suggestion
        clientId: ${AUTH_OIDC_CLIENT_ID} <2>
```

>         clientSecret: ${AUTH_OIDC_CLIENT_SECRET}

Reviewer: ```suggestion
        clientSecret: ${AUTH_OIDC_CLIENT_SECRET} <2>
```

> signInPage: oidc

Reviewer: ```suggestion
signInPage: oidc <3>
```

----

> `environment: production`::

Reviewer: ```suggestion

```

> Mark the environment as `production` to hide the Guest login in the {product-short} home page.

Reviewer: ```suggestion
<1> Mark the environment as `production` to hide the Guest login in the {product-short} home page.
```

> `metadataUrl`, `clientId`, `clientSecret`::

Reviewer: ```suggestion

```

> To configure the OIDC provider with your secrets.

Reviewer: ```suggestion
<2> To configure the OIDC provider with your secrets.
```

> `sigInPage: oidc`::

Reviewer: ```suggestion

```

> To enable the OIDC provider as default sign-in provider.

Reviewer: ```suggestion
<3> To enable the OIDC provider as the default sign-in provider.
```

> Optional: Consider adding following optional fields:

Reviewer: ```suggestion
Optional: Consider adding the following optional fields:
```

`dangerouslyAllowSignInWithoutUserInCatalog: true`::
+
--
To enable authentication without requiring to provision users in the {product-short} software catalog.
+
WARNING: Use this option to explore {product-short} features, but do not use it in production.
+

> .`app-config-rhdh.yaml` fragment with optional field to allow authenticating users absent from the software catalog

Reviewer: ```suggestion
.Example `app-config-rhdh.yaml` fragment with an optional field to allow authenticating users absent from the software catalog
```

[source,yaml]
----
auth:
  environment: production
  providers:
    oidc:
      production:
        metadataUrl: ${AUTH_OIDC_METADATA_URL}
        clientId: ${AUTH_OIDC_CLIENT_ID}
        clientSecret: ${AUTH_OIDC_CLIENT_SECRET}
signInPage: oidc
dangerouslyAllowSignInWithoutUserInCatalog: true
----
--

`callbackUrl`::
--
RHSSO callback URL.

> .`app-config-rhdh.yaml` fragment with optional `callbackURL` field

Reviewer: ```suggestion
.Example `app-config-rhdh.yaml` fragment with optional `callbackURL` field
```

[source,yaml]
----
auth:
  providers:
    oidc:
      production:
        callbackUrl: ${AUTH_OIDC_CALLBACK_URL}
----
--

`tokenEndpointAuthMethod`::
--
Token endpoint authentication method.

> .`app-config-rhdh.yaml` fragment with optional `tokenEndpointAuthMethod` field

Reviewer: ```suggestion
.Example `app-config-rhdh.yaml` fragment with optional `tokenEndpointAuthMethod` field
```

[source,yaml]
----
auth:
  providers:
    oidc:
      production:
        tokenEndpointAuthMethod: ${AUTH_OIDC_TOKEN_ENDPOINT_METHOD}
----
--

`tokenSignedResponseAlg`::
--
Token signed response algorithm.

> .`app-config-rhdh.yaml` fragment with optional `tokenSignedResponseAlg` field

Reviewer: ```suggestion
.Example `app-config-rhdh.yaml` fragment with optional `tokenSignedResponseAlg` field
```

[source,yaml]
----
auth:
  providers:
    oidc:
      production:
        tokenSignedResponseAlg: ${AUTH_OIDC_SIGNED_RESPONSE_ALG}
----
--

`scope`::
--
RHSSO scope.

> .`app-config-rhdh.yaml` fragment with optional `scope` field

Reviewer: ```suggestion
.Example `app-config-rhdh.yaml` fragment with optional `scope` field
```

[source,yaml]
----
auth:
  providers:
    oidc:
      production:
        scope: ${AUTH_OIDC_SCOPE}
----
--

`signIn.resolvers`::
--

> Declarative resolvers to override the default resolver: `emailLocalPartMatchingUserEntityName`.

Reviewer: ```suggestion
Declarative resolvers can be used to override the default resolver: emailLocalPartMatchingUserEntityName. The authentication provider attempts each sign-in resolver in sequence until one succeeds and fails if none do.
```

The authentication provider tries each sign-in resolver until it succeeds, and fails if none succeed.

> .`app-config-rhdh.yaml` fragment with optional `callbackURL` field

Reviewer: ```suggestion
.Example `app-config-rhdh.yaml` fragment with optional `callbackURL` field
```
